### PR TITLE
Decode and Encode hex

### PR DESCRIPTION
**Pull request by uaeio**
# Description

I need a command that will transform hex string into bytes and into other direction.

I've implemented `decode hex` command and `encode hex` command. (Based on `encode base64` and `decode base64` commands
# User-Facing Changes

```
> '010203' | decode hex
0x[01 02 03]
```

and 

```
> 0x[01 02 0a] | encode hex
'01020A'
```

### CODE REVIEW

File: crates/nu-command/src/strings/encode_decode/hex.rs [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+fn hex_digit(num: u8) -> char {
+    match num {
+        0..=9 => (num + b'0') as char,
+        10..=15 => (num - 10 + b'A') as char,
+        _ => unreachable!(),
+    }
+}
+
+fn hex_encode(bytes: &[u8]) -> String {
+    let mut res = String::with_capacity(bytes.len() * 2);
```
Comment (sholderbach) [inline crates/nu-command/src/strings/encode_decode/hex.rs:27]: Nice!

File: crates/nu-command/src/strings/encode_decode/hex.rs [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+fn hex_decode(value: String) -> Result<Vec<u8>, ()> {
+    let mut res = Vec::with_capacity(value.len() / 2);
+    let mut chars = value.chars();
+    while let Some(c) = chars.next() {
+        let c = c.to_digit(16).ok_or(())?;
+        let c2 = chars.next().and_then(|c| c.to_digit(16)).ok_or(())?;
+        res.push((c * 16 + c2).try_into().map_err(|_| ())?);
+    }
+    Ok(res)
+}
```
Comment (sholderbach) [inline crates/nu-command/src/strings/encode_decode/hex.rs:16]: **Nitpicking alert**

From an API perspective this doesn't have to take the `String` and could work with `&str`

Would be curious how this compares to [`u8::from_str_radix`](https://doc.rust-lang.org/core/primitive.u8.html#method.from_str_radix).

See the version in the parser:
https://github.com/nushell/nushell/blob/79d07358643193d411b85b279644d3d1157a1161/crates/nu-parser/src/parser.rs#L1397-L1406

**Review by sholderbach — COMMENTED** [review]
Thanks for the thorough implementation! I really appreciate the effort to have helpers in a separate file and digestibly-sized functions.

As we are now in the final freeze period before the upcoming release on Tuesday we will have to wait with landing.

From my time playing with those commands I might have a few suggestions how the error messages could improve or the decode command could be made more tolerant:

1) explicitly state if the error was due to forbidden characters
```
/home/stefan/nushell〉'01020G' | decode hex
Error:
  × value could not be hex decoded
   ╭─[entry #1:1:1]
 1 │ '01020G' | decode hex
   ·            ─────┬────
   ·                 ╰── invalid hex input
   ╰────
```
2) Right now it requires an even number of digits to make it unambiguous how the bytes are segmented. Maybe worth adding an error message if the number of nibbles is invalid.
```
/home/stefan/nushell〉'01020' | decode hex
Error:
  × value could not be hex decoded
   ╭─[entry #2:1:1]
 1 │ '01020' | decode hex
   ·           ─────┬────
   ·                ╰── invalid hex input
   ╰────
```

3) Whitespace in the string is disallowed right now. Not sure if that should be the case. Else the error message should probably tell you the requirements.
```
/home/stefan/nushell〉'01 02 03' | decode hex
Error:
  × value could not be hex decoded
   ╭─[entry #3:1:1]
 1 │ '01 02 03' | decode hex
   ·              ─────┬────
   ·                   ╰── invalid hex input
   ╰────
```

I was initially wondering if we can share some code with the parser here that needs to be able to parse hex literals anyways, but there might be slight differences in requirements. 
https://github.com/nushell/nushell/blob/79d07358643193d411b85b279644d3d1157a1161/crates/nu-parser/src/parser.rs#L1397-L1406

**Comment by whiteand** [comment]
@sholderbach Thanks for review, I've fixed my implementation.
1. Made more descriptive error messages
2. Made performance better (removed cloning of the string
3. Added ability to skip whitespaces during decoding

**Comment by sholderbach** [comment]
Awesome, the additional error handling that directly tells you what would be valid looks great!

**Comment by sholderbach** [comment]
@uaeio / @whiteand Would you mind patching the conflict as I can not directly push to your branch? 

```patch
From f986d052772c230fcbcb83661a911d21e3a9e789 Mon Sep 17 00:00:00 2001
From: sholderbach <sholderbach@users.noreply.github.com>
Date: Wed, 22 Mar 2023 22:41:56 +0100
Subject: [PATCH] Box `ShellError`s, fix conflict from #8375

---
 .../src/strings/encode_decode/hex.rs          | 20 +++++++++----------
 1 file changed, 10 insertions(+), 10 deletions(-)

diff --git a/crates/nu-command/src/strings/encode_decode/hex.rs b/crates/nu-command/src/strings/encode_decode/hex.rs
index 81b910e4b..56d8f997c 100644
--- a/crates/nu-command/src/strings/encode_decode/hex.rs
+++ b/crates/nu-command/src/strings/encode_decode/hex.rs
@@ -108,55 +108,55 @@ fn action(
         Value::Binary { val, .. } => match hex_config.action_type {
             ActionType::Encode => Value::string(hex_encode(val.as_ref()), command_span),
             ActionType::Decode => Value::Error {
-                error: ShellError::UnsupportedInput(
+                error: Box::new(ShellError::UnsupportedInput(
                     "Binary data can only be encoded".to_string(),
                     "value originates from here".into(),
                     command_span,
                     // This line requires the Value::Error {} match above.
                     input.expect_span(),
-                ),
+                )),
             },
         },
         Value::String { val, .. } => {
             match hex_config.action_type {
                 ActionType::Encode => Value::Error {
-                    error: ShellError::UnsupportedInput(
+                    error: Box::new(ShellError::UnsupportedInput(
                         "String value can only be decoded".to_string(),
                         "value originates from here".into(),
                         command_span,
                         // This line requires the Value::Error {} match above.
                         input.expect_span(),
-                    ),
+                    )),
                 },
 
                 ActionType::Decode => match hex_decode(val.as_ref()) {
                     Ok(decoded_value) => Value::binary(decoded_value, command_span),
                     Err(HexDecodingError::InvalidLength(len)) => Value::Error {
-                        error: ShellError::GenericError(
+                        error: Box::new(ShellError::GenericError(
                             "value could not be hex decoded".to_string(),
                             format!("invalid hex input length: {len}. The length should be even"),
                             Some(command_span),
                             None,
                             Vec::new(),
-                        ),
+                        )),
                     },
                     Err(HexDecodingError::InvalidDigit(index, digit)) => Value::Error {
-                        error: ShellError::GenericError(
+                        error: Box::new(ShellError::GenericError(
                             "value could not be hex decoded".to_string(),
                             format!("invalid hex digit: '{digit}' at index {index}. Only 0-9, A-F, a-f are allowed in hex encoding"),
                             Some(command_span),
                             None,
                             Vec::new(),
-                        ),
+                        )),
                     },
                 },
             }
         }
         other => Value::Error {
-            error: ShellError::TypeMismatch {
+            error: Box::new(ShellError::TypeMismatch {
                 err_message: format!("string or binary, not {}", other.get_type()),
                 span: other.span().unwrap_or(command_span),
-            },
+            }),
         },
     }
 }
-- 
2.25.1


```

**Comment by whiteand** [comment]
> @uaeio / @whiteand Would you mind patching the conflict as I can not directly push to your branch? 

Sure, I will do it tomorrow

**Comment by uaeio** [comment]
@sholderbach Applied patch

**Comment by sholderbach** [comment]
Awesome! Let's give it a try!

**Pull request merged**